src/cargo/core/resolver/mod.rs

```diff
@@ -884,80 +884,205 @@ fn generalize_conflicting(
-                // A + B is redundant but cant be activated, as if
-                // A + B is active then A is active and we know that is not ok.
-                for (_, other) in &others {
-                    con.extend(other.iter().map(|(&id, re)| (id, re.clone())));
+                if (our_activation_key.map_or(false, |our| {
+                    other.summary.package_id().as_activations_key() == our
+                }) || our_link.map_or(false, |_| other.summary.links() == our_link))
+                    && !our_candidates.contains(&other.summary.package_id())
+                {
+                    continue;
                 }
-                // Now that we have this combined conflict, we can do a substitution:
-                // A dep is equivalent to one of the things it can resolve to.
-                // So we can remove all the things that it resolves to and replace with the parent.
-                for (other_id, _) in &others {
-                    con.remove(other_id);
+
+                // ok so we dont have a semver nor a links problem with `critical_parents_dep`
+                // getting set to `other`, maybe it still wont work do to one of its dependencies.
+                if let Some(conflicting) = generalize_children_conflicting(
+                    cx,
+                    backtrack_critical_age,
+                    registry,
+                    past_conflicting_activations,
+                    dep,
+                    &other.summary,
+                    *critical_parent,
+                ) {
+                    con.extend(conflicting.into_iter());
+                    continue;
                 }
-                con.insert(*critical_parent, backtrack_critical_reason);
-
-                if cfg!(debug_assertions) {
-                    // the entire point is to find an older conflict, so let's make sure we did
-                    let new_age = con
-                        .keys()
-                        .map(|&c| cx.is_active(c).expect("not currently active!?"))
-                        .max()
-                        .unwrap();
-                    assert!(
-                        new_age < backtrack_critical_age,
-                        "new_age {} < backtrack_critical_age {}",
-                        new_age,
-                        backtrack_critical_age
+
+                if let Some(conflicting) = past_conflicting_activations.find(
+                    dep,
+                    &|id| {
+                        if id == other.summary.package_id() {
+                            // we are imagining that we used other instead
+                            Some(backtrack_critical_age)
+                        } else {
+                            cx.is_active(id).filter(|&age|
+                                        // we only care about things that are older then critical_age
+                                        age < backtrack_critical_age)
```

Comment:
Maybe in need of some `rustfmt` here?

Reply:
I have fmt on save, but it seems to be some kind of fmt bug. If the comment is in the callback fmt will not move it.